[wasm] wasi: Enable library tests on CI

radical opened the pull request:
- Adds tasks/targets to run a wasi console sample, and tests on CI - for linux, and windows
- Only few tests are enabled by default
- Other tests fail due to GC issues, missing fs/network access etc
- Also, adds `runtime-wasm-optional` pipeline which will have the wasi build with all the tests enabled, and will be RED. This is disabled right now.

Some differences from browser/wasm case:

```
    [wasi] Use the real entrypoint even with async Main

    When the main assembly has a async main, roslyn generates a wrapper
    method which invokes the async main. And in case of wasm that
    async-main method is invoked so it can yield to the host.

    But in case of wasi we invoke the actual entrypoint.
```

Fixes https://github.com/dotnet/runtime/issues/72641 .

pavelsavara commented:
I would like to see this merged so that I could start looking at the tests. How can I help ?

radical commented:
update: I'm cleaning this up now. I have to check the various runtime/native changes/cmake changes to make sure that they are still needed, after the earlier wasi PR being merged. I will open this up for review in a few days.

radical committed:
Add tasks from https://github.com/SteveSandersonMS/dotnet-wasi-sdk

radical committed:
build changes

radical committed:
add test build changes

radical committed:
test fixes

radical committed:
add more wasi build files

radical committed:
fix wasi sample

radical committed:
add missing files

radical committed:
disable non-wasi/wasm builds

pavelsavara reviewed (commented):
good work!

This is large PR and I made many comments, but I don't insist on any of them. 
I would like to merge it soon and fix things incrementally on top of this in further PRs.

radical commented on eng/testing/WasiRunnerTemplate.sh:
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env bash
```
Yep. I will add the windows bits after this PR.

radical commented on eng/testing/tests.browser.targets:
```diff
@@ -0,0 +1,344 @@
+<Project TreatAsLocalProperty="ArchiveTests">
```
There are bits that I'm extracting from this to the base targets file. But this is very much WIP, so tests passing is the check for the changes right now!

radical commented on src/tasks/WasmAppBuilder/wasi/EmitWasmBundleObjectFile.cs:
```diff
@@ -0,0 +1,256 @@
+
+namespace Microsoft.WebAssembly.Build.Tasks;
+
+public class EmitWasmBundleObjectFile : Microsoft.Build.Utilities.Task, ICancelableTask
```
We can remove this, or make it optional in future. Right now, it will unblock this PR.

radical commented:
I've read the feedback on all the tests targets, and wasi targets files, but ignoring them for now, because they are very much in WIP. Most of them are copies of the existing wasm files, and will get pruned in future PRs.

radical commented on src/libraries/sendtohelix-wasm.targets:
```diff
@@ -38,6 +38,7 @@
     <BuildHelixWorkItemsDependsOn>$(BuildHelixWorkItemsDependsOn);StageEmSdkForHelix;PrepareForBuildHelixWorkItems_Wasm</BuildHelixWorkItemsDependsOn>
     <IncludeHelixCorrelationPayload>false</IncludeHelixCorrelationPayload>
     <EnableDefaultBuildHelixWorkItems>false</EnableDefaultBuildHelixWorkItems>
+    <UseTestEnvFile>false</UseTestEnvFile>
```
This was part of something I was trying. It should be able to drop this now.

radical committed:
address feedback from @pavelsavara

radical committed:
Merge remote-tracking branch 'origin/main' into wasi_tests

radical committed:
cleanup

radical committed:
restore ci builds

radical committed:
fix the optional build

radical committed:
Address review feedback from @pavelsavara

radical committed:
Fix analyzer errors

radical committed:
Address review feedback from @pavelsavara

radical commented:
Thank you for the detailed feedback, @pavelsavara! I have updated the PR, and added some more details in the PR description.

radical commented on src/mono/wasi/mono-include/driver.h:
```diff
@@ -9,6 +9,7 @@ void mono_wasm_load_runtime (const char *unused, int debug_level);
 int mono_wasm_add_assembly (const char *name, const unsigned char *data, unsigned int size);
 MonoAssembly* mono_wasm_assembly_load(const char *name);
 MonoMethod* mono_wasm_assembly_get_entry_point (MonoAssembly *assembly);
```
This was my original commit message which kinda got lost in re-arranging the commits!
```
    [wasi] Use the real entrypoint even with async Main

    When the main assembly has a async main, roslyn generates a wrapper
    method which invokes the async main. And in case of wasm that
    async-main method is invoked so it can yield to the host.

    But in case of wasi we invoke the actual entrypoint.
```

radical commented on src/mono/mono/utils/mono-threads-wasm.c:
```diff
@@ -48,19 +52,22 @@ uintptr_t get_wasm_stack_high(void);
-wasm_get_stack_size (void)
+wasm_get_stack_base (void)
 {
+    return ((uintptr_t)&__heap_base) - wasm_get_stack_size();
```
This was working earlier, but I kept it as-is for this PR. This and the gc will need fixes.
But this is only affecting wasi, so as long as CI is green, we can merge it in this PR. And make changes in follow up.

radical committed:
gen-buildsys.cmd: fix wasi sdk path on windows

radical committed:
Always set arch=wasm for OS=browser, and OS=wasi

radical committed:
[wasi] Fix wasi/windows build

radical committed:
cleanup

radical committed:
Add wasi windows job

radical committed:
Merge remote-tracking branch 'origin/main' into wasi_tests

radical committed:
Fix building sample

radical committed:
Fixes for running wasi sample, and library tests - System.Buffers.Tests passes

radical committed:
fix wasm tests

radical committed:
fix wasm tests

radical committed:
Merge remote-tracking branch 'origin/main' into wasi_tests

radical commented:
I'm testing enabling tests now windows in https://github.com/dotnet/runtime/pull/81452 .

pavelsavara committed:
fix sample

pavelsavara committed:
runner script for windows

radical committed:
forgot to add WasiRunnerTemplate.cmd

pavelsavara committed:
fix sample again

radical commented:
/azp run runtime-wasm

radical committed:
[wasm] fix wasi tests on windows/CI

radical committed:
[wasm] Add wasi tests for CI/win

pavelsavara commented on src/mono/wasi/wasi.proj:
```diff
@@ -6,8 +6,9 @@
   <PropertyGroup>
-    <ICULibDir Condition="'$(MonoWasmThreads)' != 'true'">$([MSBuild]::NormalizeDirectory('$(PkgMicrosoft_NETCore_Runtime_ICU_Transport)', 'runtimes', 'wasi-wasm', 'native', 'lib'))</ICULibDir>
-    <ICULibDir Condition="'$(MonoWasmThreads)' == 'true'">$([MSBuild]::NormalizeDirectory('$(PkgMicrosoft_NETCore_Runtime_ICU_Transport)', 'runtimes', 'wasi-wasm-threads', 'native', 'lib'))</ICULibDir>
+    <!-- TODOWASI: icu support -->
```
Could we enable this again before merge or this breaks stuff ? 
cc @mkhamoyan

radical replied:
I'll try.

radical replied:
With this uncommented, a test build fails with:
```
wasm-ld : error : /workspaces/runtime/artifacts/bin/microsoft.netcore.app.runtime.wasi-wasm/Release/runtimes/wasi-wasm/native/libicudata.a: archive has no index; run ranlib to add one [/workspaces/runtime/src/libraries/System.Buffers/tests/System.Buffers.Tests.csproj]
wasm-ld : error : /workspaces/runtime/artifacts/bin/microsoft.netcore.app.runtime.wasi-wasm/Release/runtimes/wasi-wasm/native/libicui18n.a: archive has no index; run ranlib to add one [/workspaces/runtime/src/libraries/System.Buffers/tests/System.Buffers.Tests.csproj]
wasm-ld : error : /workspaces/runtime/artifacts/bin/microsoft.netcore.app.runtime.wasi-wasm/Release/runtimes/wasi-wasm/native/libicuuc.a: archive has no index; run ranlib to add one [/workspaces/runtime/src/libraries/System.Buffers/tests/System.Buffers.Tests.csproj]
```

mkhamoyan replied:
I'l take a look.

pavelsavara replied:
https://github.com/dotnet/runtime/issues/81560

radical committed:
Merge remote-tracking branch 'origin/main' into wasi_tests

pavelsavara committed:
fix stack range again

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
* stack first

 * also for relink

pavelsavara commented on src/mono/mono/metadata/sgen-mono.c:
```diff
@@ -2293,7 +2293,7 @@ sgen_client_scan_thread_data (void *start_nursery, void *end_nursery, gboolean p
 	scan_area_arg_start = start_nursery;
 	scan_area_arg_end = end_nursery;
-#ifdef HOST_WASM
+#if defined(HOST_WASM) || defined(HOST_WASI)
```
should be rather `HOST_BROWSER` ?

pavelsavara committed:
- back to assembly in .S file. extern doesn't work is zero

 - asserts about order of memory segments
 - further unified re-link and get rig of --stack-first

radical commented:
Wasi tests are failing on `	g_assert (heap_base - stack_size_pad == data_end);`
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-81052-merge-b56c232ac68c49339e/Microsoft.XmlSerializer.Generator.Tests/1/console.e72d0849.log?helixlogtype=result

radical committed:
[wasm] wasi: Fix assert, and compute stack size at run time

the current memory layout:

```
  /*
   * | -- increasing address ---> |
   * | data (data_end)| stack |(heap_base) heap |
   */
```

.. based on which we can calculate a reasonable stack size.

Tests were failing with the earlier code on the following assert:

```c
size_t stack_size_pad = wasm_get_stack_size() + 12; // there is padding
g_assert (heap_base - stack_size_pad == data_end);
```

radical committed:
fix typos in makefile

radical committed:
Run System.Collections.Tests also, which would test the oom case

radical commented:
/azp run runtime-wasm
